sentry_sdk/tracing_utils.py

```diff
@@ -160,6 +191,97 @@ def maybe_create_breadcrumbs_from_span(hub, span):
         )
 
 
+def extract_sentrytrace_data(header):
+    # type: (Optional[str]) -> typing.Mapping[str, Union[Optional[str], Optional[bool]]]
+
+    """
+    Given a `sentry-trace` header string, return a dictionary of data.
+    """
+    trace_id = parent_span_id = parent_sampled = None
+
+    if header:
+        if header.startswith("00-") and header.endswith("-00"):
+            header = header[3:-3]
+
+        match = SENTRY_TRACE_REGEX.match(header)
+
+        if match:
+            trace_id, parent_span_id, sampled_str = match.groups()
+
+            if trace_id:
+                trace_id = "{:032x}".format(int(trace_id, 16))
+            if parent_span_id:
+                parent_span_id = "{:016x}".format(int(parent_span_id, 16))
+            if sampled_str:
+                parent_sampled = sampled_str != "0"
+
+    return {
+        "trace_id": trace_id,
+        "parent_span_id": parent_span_id,
+        "parent_sampled": parent_sampled,
+    }
+
+
+def extract_tracestate_data(header):
+    # type: (Optional[str]) -> typing.Mapping[str, Optional[str]]
+    """
+    Extracts the sentry tracestate value and any third-party data from the given
+    tracestate header, returning a dictionary of data.
+    """
+    sentry_value = third_party = None
+
+    if header:
+
+        match = TRACESTATE_HEADER_REGEX.match(header)
+
+        if match:
+            before, sentry_value, after = match.groups()
+            if before or after:
+                # filter out empty strings, and make sure there aren't too many
+                # commas between them
+                third_party = ",".join(
+                    [value.strip(",") for value in [before, after] if value != ""]
+                )
+
+        else:
+            # if the header is malformed, at least try to grab sentry's part, if any
+            sentry_value_match = TRACESTATE_SENTRY_VALUE_REGEX.search(header)
+            if sentry_value_match:
+                sentry_value = sentry_value_match.group(1)
```

Comment:
I think you might have a bug here:

```
>>> x =TRACESTATE_HEADER_REGEX.match("abc=11,sentry=2223,def=33")
>>> x.groups()
('abc=11,', '2223', '2223', None, ',def=33') # cool
>>> x =TRACESTATE_HEADER_REGEX.match("abc=11,def=33")
>>> x is None
True  # not cool 
>>> x =TRACESTATE_HEADER_REGEX.match("abc=11,,def=33")
>>> x.groups()
('abc=11,', None, None, None, ',def=33') # not cool

```
You don't recognize the ok trace state `abc=11,def=33` and then you will fall back on trying to extract the sentry tracestate (throwing away the third party trace state).

So the problem is that you expect to have somewhere a `sentry=xxx` entry, if not you expect a sequence of (xxx=yyy,)* followed by (,xxx=yyy) so when sentry=xxx is missing you expect somewhere to either have `,,` or to start or end with a `,`

Reply:
More general I think the problem you are trying to solve can't be solved the way you are trying. That is how I got suspicious and tried to understand how it is working. Ignoring for a second the separator `,` you are effectively trying to match something like this (.*)(a*)(.*) where I replaced `(xxx=yyy)*` with `.*` and `(sentry=yyy)*` with `a*`. The problem is that `a*` is a special case of `.*` and since * is greedy (.*)(a*)(.*) it is always going to match everything in the first group and will leave the last two groups empty. In your case that doesn't happen because of the way you match `,` but that causes other problems.

The way I would approach it is to look for `sentry=yyy` and then concatenate what is before and after taking care of removing one `,` if needed. That would perhaps preserve some invalid third party states but there is no real harm in that (it is not our concern and as long as we don't damage anything else there is no harm done).

If you want to be more rigorous you can match all `xxx=yyyy` and manually look among all matches for 'sentry=yyy' and put it in front.

Reply:
Yup - I found this out once I started writing tests for it. I've reworked the entire function.